vectorflow.math: Enable flushing denormals to zero

CyberShadow opened the pull request:
I noticed that when I built the mnist example with LDC and optimizations enabled, the displayed loss was NaN and the network wasn't being trained (classification error remained at 0.9). I tracked this down to some vector operations on denormal floats resulting in NaNs, which then propagated all over the place.

The fix is simple - enable the FPU mode to flush denormals to zero.

This gets the mnist example working properly on my machine when built with LDC and optimizations.

CyberShadow committed:
vectorflow.math: Enable flushing denormals to zero

CyberShadow committed:
Make ct_msg a mixin template

Using a regular function to print a compile-time message still causes
the compiler to emit an empty function in the object file; a
compile-time construct is better suited for compile-time side effects.

Use a mixin template to allow usage both inside and outside function
scopes.

Removing the unnecessary static constructor in vectorflow.neuralnet
fixes the cyclic dependency issue with vectorflow.math in non-release
builds.

CyberShadow committed:
vectorflow.math: Work around DMD bug 16098

align(...) is not respected for stack variables.

rostyboost commented:
Thanks @CyberShadow ! This is interesting. Can you provide the flags and version of LDC for which you noticed the issue in the first place, as well your running arch and the # of cores & epochs used in training? Lock-Free async SGD will definitely produce this behavior if run for too long (or too many races across threads on a dense model generating a high level of weight updates contention).
The FTZ mitigation in these pathological cases is a good idea, even though one would typically tweak the learning parameters better.

I've been meaning to do something like this for other reasons (mostly performance, but needs some benchmarking) but I have always wondered about the portability aspect. Do you have any insights on this? Current CI isn't set up to check for ARM32 compat for example.
While we're at it, we should also set `DAZ`. Large-scale ML doesn't require anything near this when it comes to precision.

CyberShadow committed:
vectorflow.math: Set DAZ (denormals are zero) bit, too

CyberShadow committed:
vectorflow.math: Don't attempt to use fxsave/mxcsr on non-x86 archs

CyberShadow commented:
Hi!

> Can you provide the flags and version of LDC for which you noticed the issue in the first place, as well your running arch and the # of cores & epochs used in training?

I'm using LDC 1.13.0 from the Arch Linux package, and can reproduce this problem when building / running the `mnist` example unmodified (so, with 4 threads and 15 passes), built with `ldmd -O`.

> I've been meaning to do something like this for other reasons (mostly performance, but needs some benchmarking) but I have always wondered about the portability aspect. Do you have any insights on this?

Yes, unfortunately it doesn't look like there is a portable way to set these flags in C, much less D. The best options I found was to use raw assembler (as here), or through some platform-specific C headers (e.g. `FE_DFL_DISABLE_SSE_DENORMS_ENV` or `_MM_SET_FLUSH_ZERO_MODE`) , which are not available in D.

> Current CI isn't set up to check for ARM32 compat for example.

Sorry, I don't know how to set this flag for ARM CPUs.

I added a commit to only enable this code on x86 architectures.

>  While we're at it, we should also set `DAZ`. Large-scale ML doesn't require anything near this when it comes to precision.

OK, added a commit doing so.

rostyboost commented:
Thanks @CyberShadow . Will try to run some benchmarks with this patch over the coming week.

rostyboost commented:
Finally took time to look at this one. Looks great, I'm merging it now. Thanks for your contribution @CyberShadow !
